docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+you *must* log it (if it needs to be logged at all).
+
+For example:
+
+```rust
+fn read_motd_file() -> std::io::Result<String> {
+    let mut f = File::open("/etc/motd")?;
+    let mut result = String::new();
+    f.read_to_string(&mut result)?;
+    result
+}
+```
+
+Opening or reading the file could fail, but there is no need to log
+the error here. The function merely propagates the error to the
+caller, and it is up to the caller to log the error or propagate it
+further, if the failure is not expected. But if, for example, it is
+normal that the "/etc/motd" file doesn't exist, the caller can choose
+to silently ignore the error, or log it in INFO or DEBUG level:
+
+```rust
+fn get_message_of_the_day() -> String {
+    // Get the motd from /etc/motd, or return the default proverb
+    match read_motd_file() {
+        Ok(motd) => motd,
+        Err(err)  => {
+            // It's normal that /etc/motd doesn't exist, but if we fail to
+            // read it for some other reason, that's unexpected. The message
+            // of the day isn't very important though, so we just WARN and
+            // continue with the default in any case.
+            if err.kind() != std::io::ErrorKind::NotFound {
+                 tracing::warn!("could not read \"/etc/motd\": {err:?}");
+            }
+            "An old error is always more popular than a new truth. - German proverb"
+        }
+    }
+}
+```
+
+## Error types
+
+We use the `anyhow` crate widely. It contains many convenient macros
+like `bail!` and `ensure!` to construct and return errors, and to
+propagate many kinds of low-level errors, wrapped in `anyhow::Error`.
+
+A downside of `anyhow::Error` is that the caller cannot distinguish
+between different error cases. Most errors are propagated all the way
+to the mgmt API handler function, or the main loop that handles a
+connection with the compute node, and they are all handled the same
+way: the error is logged and returned the error to the client as
+an HTTP or libpq error.
+
+But in some cases, we need to distinguish between errors and handle
+them differently. For example, attaching a tenant to the pageserver
+could fail either because the tenant has already been attached, or
+because we could not load its metadata from cloud storage. The first
+case is more or less expected. The console sends the Attach request to
+the pageserver, and the pageserver completes the operation, but the
+network connection might be lost before the console receives the
+response. The console will retry the operation in that case, but the
+tenant has already been attached. It is important that the pagserver
+responds with the HTTP 403 Already Exists error in that case, rather
+than a generic HTTP 500 Internal Server Error.
+
+If you need to distinguish between different kinds of errors, create a
+new `Error` type. The `thiserror` create is useful for that. But in
+most cases `anyhow::Error` is good enough.
```

Comment:
It seems that I'm in minority here. But I'll raise my hand anyway :)

To me error handling with anyhow is broken for anything other than simple CLIs which just bubble up errors to the user without trying to inspect them. 

The ideal approach to me is that every module defines its error enum with `thiserror::Error` (it just saves boilerplate on writing impls manually). It has the benefit that anyone who sees this module for the first time can see what are the possible errors. Listing error variants requires to put more thought into it compared to just a bunch of `bail!("A is broken")` calls. Better names for errors lead to better answers to "How I can handle this error?" type of questions. 

Nesting anyhow::Error types in other enum has following downsides:
* When there are many anyhow::Error variants `From` doesnt work anymore which leads to boilerplate since `?` is not available.
* If you need to detalize some anyhow::Error into new variants then you have to change every caller of a function returning such an error. With thiserror you need only add new variant and change the places which actually handle the error

This approach is more radical and requires ground up design for error types which is not an easy task